### PR TITLE
[big merge][student-teacher] Sophiex/dev/ssl losses 1043

### LINKED ISSUE #1043: Student-Teacher Loss calculator
### Is your feature request related to a problem? Please describe.

A general Loss Calculator that can implement a variety of contrastive student-teacher style representation learning techinques such as BYOL, JEPA, DINO, etc

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

[DRAFT] PR for introducing the losses for SSL student-teacher latent losses. This PR will rely on both the abstract loss calculator #1178  as well as the abstract target/aux class #1179

The idea is to get early feedback and notice issues my making code more concrete

## Issue Number

Closes #1043 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/ssl_losses_utils.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
Comment (clessig) [inline src/weathergen/train/ssl_losses_utils.py:3]: This file should go to . They need to be torch.nn.modules because this are NNs, even if they are not necessarily themselves trained. I think `ssl_target_processing.py` (since you probably still don't like `ssl_target_predictors.py`)

File: src/weathergen/train/ssl_losses_utils.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+def lossfunc(t, s, temp):
```
Comment (clessig) [inline src/weathergen/train/ssl_losses_utils.py:23]: The name is not very descriptive :) Maybe `latent_logit_loss.py`? JEPA uses MAE (and one could conceivably replace by MSE) which are already implemented in loss.py. Ideally we could reuse what is there.

File: src/weathergen/train/ssl_losses_utils.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+            Q /= K
+
+            # normalize each column: total weight per sample must be 1/B
+            Q /= torch.sum(Q, dim=0, keepdim=True)
+            Q /= B
+
+        Q *= B  # the columns must sum to 1 so that Q is an assignment
+        return Q.t()
+
+    # def forward(self, student_patch_tokens, teacher_patch_tokens, student_masks_flat):
```
Comment (clessig) [inline src/weathergen/train/ssl_losses_utils.py]: Can we remove the stale code? What does it implement?

File: src/weathergen/train/ssl_losses_utils.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            )
+            self.update_center(teacher_output)
+            return processed_teacher_output
+        elif self.teacher_style == "sinkhorn_knopp":
+            return self.sinkhorn_knopp_teacher(teacher_output, self.teacher_temp)
+        else:
+            # this code should never be reached, see assert in __init__
+            return teacher_output
+
+    # def forward(self, student_output_list, teacher_out_softmaxed_centered_list):
```
Comment (clessig) [inline src/weathergen/train/ssl_losses_utils.py]: Remove

File: src/weathergen/train/ssl_losses_utils.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    https://github.com/facebookresearch/dinov2/tree/main
+
+    Needs to be nn.Module because of the registered_buffer, it means we should have a forward
+    function, previously was the softmax computation, maybe we can make it the
+    softmax_center_teacher, etc
+    """
+
+    def __init__(
+        self,
+        patch_out_dim,
```
Comment (clessig) [inline src/weathergen/train/ssl_losses_utils.py:39]: Would it be better to take a dict as arg if we potentially want to implement *TargetProcessing that requires different args.

**Review by clessig — COMMENTED** [review]
Didn't look through the actual computations line by line since it seems this copy-paste from the reference code?

File: src/weathergen/train/ssl_losses_utils.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/ssl_losses_utils.py]: the stale code is there for reference because it needs to go to the loss calculator later

I will do all the clean-up once we are much closer to actually merging :)

File: src/weathergen/train/ssl_losses_utils.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+import torch.nn.functional as F
+
+
+def lossfunc(t, s, temp):
+    return torch.sum(t * F.log_softmax(s / temp, dim=-1), dim=-1)
+
+
+class iBOTPatchTargetProcessing(nn.Module):
+    """
+    Code taken and adapted from the official DINOv2 implementation
```
Comment (tjhunter) [inline src/weathergen/train/ssl_losses_utils.py:29]: could you point to the actual file? it will help to understand what got copied exactly:
https://github.com/facebookresearch/dinov2/blob/main/dinov2/loss/ibot_patch_loss.py

Also, based on the license, we will need to put in the README of the project that some portion of WG is Copyright (c) Meta Platforms, Inc. and affiliates.

File: src/weathergen/train/ssl_losses_utils.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+
+            self.center = self.center * self.center_momentum + _t * (1 - self.center_momentum)
+
+            self.updated = True
+
+
+class DINOTargetProcessing(nn.Module):
+    """
+    Code taken and adapted from the official DINOv2 implementation
+    https://github.com/facebookresearch/dinov2/tree/main
```
Comment (tjhunter) [inline src/weathergen/train/ssl_losses_utils.py:152]: same comment here

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -313,8 +315,19 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
...
-            meta_ema_model = self.init_model_and_shard(cf, devices)[0]
+            meta_ema_model = self.init_model_and_shard(cf, "student", devices)[0]
+            self.ema_model = EMAModel(
+                self.model,
+                meta_ema_model,
+                halflife_steps=cf.get("ema_halflife_in_thousands", 1e-3),
+                rampup_ratio=cf.get("ema_ramp_up_ratio", 0.09),
+                is_model_sharded=(cf.with_ddp and cf.with_fsdp),
+            )
+        elif cf["training_mode"] == "student-teacher":
```
Comment (tjhunter) [inline src/weathergen/train/trainer.py]: small comment: prefer in general `cf.get(...)` for backward compatibilty

File: src/weathergen/train/target_and_aux_ssl_teacher.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        for loss_name, target_module in self.postprocess_targets.items():
+            with torch.no_grad():
+                targets[loss_name] = None  # target_module(targets["loss_name"])
+        return targets, None
+
+
+def get_target_postprocessing(target_losses: list[str], **kwargs):
+    return_dict = {}
+    for loss_name in target_losses:
+        if loss_name == "iBOT":
```
Comment (tjhunter) [inline src/weathergen/train/target_and_aux_ssl_teacher.py:59]: FYI, python also has a `match loss_name: case "iBot"` syntax.

File: src/weathergen/train/target_and_aux_ssl_teacher.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            return_dict[loss_name] = DINOTargetProcessing(
+                out_dim=kwargs["dino_out_dim"],
+                center_momentum=kwargs["center_momentum"],
+                student_temp=kwargs["student_temp"],
+                teacher_style=kwargs["teacher_style"],
+            )
+        elif loss_name == "JEPA":
+            return_dict[loss_name] = JEPATargetProcessing()
+        else:
+            # We skip losses that are not handled by the EMATeacher
```
Comment (tjhunter) [inline src/weathergen/train/target_and_aux_ssl_teacher.py:77]: I would abort to make it explicit that some of the config is not valid. It is more likely to be a bug than a conscious decision.

**Review by tjhunter — COMMENTED** [review]
some initial comments. looking forward to seeing it in action. 

High level comment: the current teacher-student framework wraps the whole model. Do we want that? I always thought it would be applied more locally up to the global assimilation engine. It would simplify future interactions with the diffusion part in the forecasting engine.

File: src/weathergen/train/ssl_losses_utils.py (hunk 6 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/ssl_losses_utils.py:29]: Let's actually discuss that last point with Christian, we may want to avoid that?

File: src/weathergen/train/target_and_aux_ssl_teacher.py (hunk 9 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_ssl_teacher.py:59]: when was this introduced? not sure how readable it is more people :/

File: src/weathergen/train/target_and_aux_ssl_teacher.py (hunk 10 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/target_and_aux_ssl_teacher.py:77]: No we need it like this for flexibility, because eg a physical space reconstruction loss wouldn't be handled by this Teacher, but are valid and would be in this list :)

File: src/weathergen/train/loss_modules/loss_module_ssl.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                name, preds, [view for view in view_metadata if view.parent_view_id is not None]
+            )
+            targets_for_loss = gather_targets_for_loss(
+                name,
+                targets,
+                [view for view in view_metadata if view.parent_view_id is None],
+            )
+            loss_value = loss_fn(preds_for_loss, targets_for_loss).mean()
+            loss += weight * loss_value
+            losses_all[name] = loss_value.item()
```
Comment (Jubeku) [inline src/weathergen/train/loss_modules/loss_module_ssl.py:72]: Hi Sophie, it seems you are using `name` in `losses_all[name]` for your loss type for now (e.g. `name="DINO"`), while we used to use it for the stream name. Should `losses_all` be a nested directory, i.e. `losses_all[stream.name][loss_type]`?

Comment (Jubeku) [reply src/weathergen/train/loss_modules/loss_module_ssl.py:72]: Wait, `stream.name` anyways doesn't make sense for latent losses. Maybe it does make sense to allow name to be something general.

Comment (sophie-xhonneux) [reply src/weathergen/train/loss_modules/loss_module_ssl.py:72]: oh I see, I thought it would be useful to have invidual losses also stored there, I am open to suggestions!

Comment (kacpnowak) [reply src/weathergen/train/loss_modules/loss_module_ssl.py:72]: Hi Sophie and Julian, `losses_all` are being collected from all ranks and matched with streams later in [this method](https://github.com/ecmwf/WeatherGenerator/blob/0a2bdbae085b710afc2b784fc32df5ed7fc4acab/src/weathergen/train/trainer.py#L966). 

In my opinion, it would be the safest way to save latent losses separately. One could add `loss_latent_hist = []` [here](https://github.com/ecmwf/WeatherGenerator/blob/0a2bdbae085b710afc2b784fc32df5ed7fc4acab/src/weathergen/train/trainer.py#L590) and use it in place of `losses_all`

Comment (Jubeku) [reply src/weathergen/train/loss_modules/loss_module_ssl.py:72]: Thanks Kacper. Yes, I am trying to consolidate this to hold across multiple loss classes

File: src/weathergen/model/model.py [hunk 12]
```diff
@@ -27,20 +30,32 @@
...
 
 
+@dataclasses.dataclass
+class ModelOutput:
+    """
+    A dataclass to encapsulate the model output and give a clear API.
+    """
+
+    physical: dict[str, torch.Tensor]
+    latent: dict[str, torch.Tensor | LatentState | DiagonalGaussianDistribution]
```
Comment (tjhunter) [inline src/weathergen/model/model.py]: <3

File: src/weathergen/train/loss_modules/__init__.py [hunk 13]
```diff
@@ -0,0 +1,5 @@
+from .loss_module_latent import LossLatent
```
Comment (tjhunter) [inline src/weathergen/train/loss_modules/__init__.py]: nit: would you mind using absolute paths? relative paths tend to break in mysterious and confusing ways

File: config/default_config.yml [hunk 14]
```diff
@@ -91,39 +92,81 @@ validate_with_ema: True
...
+      "weight": 1.0,
+      "iBOT": {'weight': 0.5, "loss_extra_args": { "student_temp": 0.1,},"out_dim": 16384, "teacher_temp": 0.1,
+          "teacher_style": "softmax_center", "center_momentum": 0.9}, 
+      "DINO": {'weight': 0.5, "loss_extra_args": { "student_temp": 0.1,}, "out_dim": 16384, "teacher_temp": 0.1,
+          "teacher_style": "softmax_center", "center_momentum": 0.9}, 
+      }
+      # "JEPA": {'weight': 5, "loss_extra_args": {}, "out_dim": 2048} }
+# Student-teacher configuration (only used when training_mode == "student_teacher")
+# TODO: adapt so that the masking or forecast config entry also sits here
+training_config:
```
Comment (kctezcan) [inline config/default_config.yml:101]: should we call these training_settings instead of _config to avoid having clashing names?

File: src/weathergen/datasets/stream_data.py [hunk 15]
```diff
@@ -131,7 +129,6 @@ def add_empty_target(self, fstep: int) -> None:
         """
 
         self.target_tokens[fstep] += [torch.tensor([], dtype=torch.int32)]
-        self.target_tokens_lens[fstep] += [torch.zeros([self.healpix_cells], dtype=torch.int32)]
         self.target_coords[fstep] += [torch.zeros((0, 105)) for _ in range(self.healpix_cells)]
```
Comment (kctezcan) [inline src/weathergen/datasets/stream_data.py]: sure about the hardcoded 105 here?

Comment (clessig) [reply src/weathergen/datasets/stream_data.py]: The 105 is there for a long time. But maybe someone can tidy this up ;)

File: config/default_config.yml (hunk 14 above)
Comment (sophie-xhonneux) [reply config/default_config.yml:101]: which name would clash?

File: src/weathergen/train/ssl_losses_utils.py (hunk 1 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/ssl_losses_utils.py:3]: You want to rename the file simply? I can do that, just want to make sure I understand correctly

Comment (clessig) [reply src/weathergen/train/ssl_losses_utils.py:3]: Yes, I would rename it, and move to weathergen/models/. Sorry, the comment was incomplete. For me, there is key functionality in this file and I wouldn't expect it in a file called utils. For me it's also mainly a model part although I see that it's close to the loss computation.

**Review by clessig — APPROVED** [review]
Awesome work! Excited to see the experiments!
